### PR TITLE
[Inductor] Expand dtype aware codegen for libdevice and tl.math ops

### PR DESCRIPTION
**Pull request by blaine-rister**
# Feature
Previously, only the codegen for `torch.sqrt` was dtype aware. This PR updates most of the `libdevice`/`tl.math` ops to support dtype-aware codegen as well. This is often necessary to get correct code when `config.triton.codegen_upcast_to_fp32=False`, as most Triton math ops do not support float16/bfloat16.

This PR enables dtype aware codegen via the `maybe_upcast_float32` decorator. This wraps `TritonOverrides` macros to upcast arguments to float32, and downcast the result back to the original dtype. The exception is for ops that return booleans, in which case we set `convert_output=False` and skip the output cast.

# Test Plan
Added CI tests for all the new ops. The list of ops to test is automatically generated based on uses of the `maybe_upcast_float32` decorator, and stored in the new `OpDtypeSupport` class. In each new test, we search the generated code for upcasts/downcasts using a regex.

Also added a unit test for `OpDtypeSupport` which checks that we have correct dtype info for ops that require upcasts. 

This PR also moves some existing tests around, to collect all the dtype aware codegen tests in one file.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: test/inductor/test_op_dtype_prop.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            "sin",
+            "cos",
+            "exp",
+            "expm1",
+            "exp2",
+            "abs",
+        ]:
+            # These ops do not support float16/bfloat16.
+            supported_dtypes = OpDtypeSupport.supported_dtypes[op_name]
+            self.assertNotIn(torch.float16, supported_dtypes)
```
Comment (blaine-rister) [inline test/inductor/test_op_dtype_prop.py:134]: TODO assert float32/float64 are supported

**Comment by blaine-rister** [comment]
Looks like the CI failures are inherited from https://github.com/pytorch/pytorch/pull/140057. Once that one lands, I'll rebase and hope for a green CI.

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -727,6 +744,39 @@ def update_on_args(self, name, args, kwargs):
...
+    This decorates tl.math/libdevice codegen functions.
+    """
+
+    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
+        # Record that this function only supports float32 and float64.
+        OpDtypeSupport.register_upcast(func, convert_output)
+
+        def wrapped(*args, **kwargs) -> str:
+            # Optionally upcast args to float32.
+            dtype = args[0].dtype
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py:759]: Some of these take in multiple inputs - can we iterate over args, kwargs ? in case args[1] is fp16, etc

**Review by eellison — APPROVED** [review]
Looks good ! i think next step is to test `test_op_dtype_propagation` with bf16, fp16 inputs and turn off upcast_to_fp32. although i think we'll need to handle the div/mod downcast as well.. we can follow up abt that, but lets land this first

**Review by arui-meta — APPROVED** [review]

File: torch/_inductor/codegen/triton.py (hunk 2 above)
Comment (blaine-rister) [reply torch/_inductor/codegen/triton.py:759]: Good point. I originally wrote this only for unary ops, but now the PR is supposed to cover binary as well. I'll add a test case for a binary op where only one input needs to be upcast.

Comment (blaine-rister) [reply torch/_inductor/codegen/triton.py]: I added the logic and a corresponding test in [77ee9a3](https://github.com/pytorch/pytorch/pull/140864/commits/77ee9a3c3c1c17e6e782e3e83c5d423092dbf2d5). I struggled to find a good test case, but the logic seems to make sense and doesn't break anything. At least for the cases I tried, it seems like inductor will insert casts at the graph level anyways so input tensors to tl.math/libdevice always end up having the same dtype by the time we get to codegen.

**Comment by blaine-rister** [comment]
Importing just to make sure this doesn't break internal tests.

**Comment by blaine-rister** [comment]
Canceling merge until I can confirm that internal tests are fixed.

**Pull request closed without merging**